lib/route/cls.c

```diff
@@ -364,6 +364,76 @@ void rtnl_cls_cache_set_tc_params(struct nl_cache *cache,
 	cache->c_iarg2 = parent;
 }
 
+/**
+ * Search classifier by interface index, parent and handle
+ * @arg cache           Classifier cache
+ * @arg ifindex         Interface index
+ * @arg parent          Parent
+ * @arg handle          Handle
+ *
+ * Searches a classifier cache previously allocated with rtnl_cls_alloc_cache()
+ * and searches for a classifier matching the interface index, parent
+ * and handle.
+ *
+ * The reference counter is incremented before returning the classifier,
+ * therefore the reference must be given back with rtnl_cls_put() after usage.
+ *
+ * @return Classifier or NULL if no match was found.
+ */
+struct rtnl_cls *rtnl_cls_get(struct nl_cache *cache, int ifindex, uint32_t parent,
+			      uint32_t handle)
+{
+	struct rtnl_cls *cls;
+
+	if (cache->c_ops != &rtnl_cls_ops)
+		return NULL;
+
+	nl_list_for_each_entry(cls, &cache->c_items, ce_list) {
+		if ((cls->c_parent == parent) && (cls->c_ifindex == ifindex)
+		    && (cls->c_handle == handle)) {
+			nl_object_get((struct nl_object *) cls);
+			return cls;
+		}
+	}
+
+	return NULL;
+}
+
+/**
+ * Search classifier by interface index, parent and priority
+ * @arg cache           Classifier cache
+ * @arg ifindex         Interface index
+ * @arg parent          Parent
+ * @arg prio            Priority
+ *
+ * Searches a classifier cache previously allocated with rtnl_cls_alloc_cache()
+ * and searches for a classifier matching the interface index, parent
+ * and prio.
+ *
+ * The reference counter is incremented before returning the classifier,
+ * therefore the reference must be given back with rtnl_cls_put() after usage.
+ *
+ * @return Classifier or NULL if no match was found.
+ */
+struct rtnl_cls *rtnl_cls_get_by_prio(struct nl_cache *cache, int ifindex,
+				      uint32_t parent, uint16_t prio)
+{
+	struct rtnl_cls *cls;
+
+	if (cache->c_ops != &rtnl_cls_ops)
+		return NULL;
+
+	nl_list_for_each_entry(cls, &cache->c_items, ce_list) {
+		if ((cls->c_parent == parent) && (cls->c_ifindex == ifindex)
+		    && (cls->c_prio == prio)) {
+			nl_object_get((struct nl_object *) cls);
```

Comment:
what do you think about having the new API not increase the ref-count?

Reply:
Good question. I've been browsing through the code and it seems like a common practice to increase the reference counter. However, I do not see any major hazards to removing _nl_obj_get()_ from the new API, since it would make things somewhat simpler: the user wouldn't have to bother calling respective _*_put()_ function for his object.
**The common scenario is this I believe:**
1. cache_alloc()
2. fetch object from cache
3. use object
4. cache_free()
Destroying the cache would invoke _nl_obj_put() -> nl_obj_free()_ chain for each object.

**There's one corner case popping out as the result of such change though:**
_1. cache_alloc()
2. fetch object from cache
3. cache_free()
4.let's use our object -> ooops, it is nullified._

Having the refcount incremented every time we fetch object from cache, would never leave us with nullified object.

Which way to go is a matter of preference. What do you think?

Reply:
> Which way to go is a matter of preference.

Yes, it's a matter of preference. I think it is wrong to increment the reference counter in this case, because the cache is not thread-safe. So that means, if we would return the object without a new reference, then nobody (no other thread) could remove the object while we have it. That means, it's not necessary, and make handling only more cumbersome, because you basically always need to return the reference.

However, much of existing code does return a reference. Consistency is also important.

Both approaches thus have up and downsides, I think I am fine with both (and with the latest version of the patch).